src/lib/comms/sol-lwm2m.c

```diff
@@ -2198,13 +2198,91 @@ sol_lwm2m_resource_init(struct sol_lwm2m_resource *resource,
         uint16_t until = i;
 
         for (i = 0; i < until; i++)
-            sol_blob_unref(resource->data[i].blob);
+            sol_blob_unref(resource->data[i].content.blob);
     }
     free(resource->data);
     va_end(ap);
     return r;
 }
 
+SOL_API int
+sol_lwm2m_resource_init_vector(struct sol_lwm2m_resource *resource,
+    uint16_t id, enum sol_lwm2m_resource_data_type data_type,
+    struct sol_vector *res_inst_ids, struct sol_vector *res_inst_values)
+{
+    uint16_t i;
+    int r = -EINVAL;
+    uint16_t resource_len = res_inst_ids->len;
+
+    if (!resource || data_type == SOL_LWM2M_RESOURCE_DATA_TYPE_NONE ||
+        !resource_len || sizeof(res_inst_ids->elem_size) != sizeof(uint16_t) ||
+        res_inst_ids->len != res_inst_values->len)
+        return -EINVAL;
+
+    LWM2M_RESOURCE_CHECK_API(resource, -EINVAL);
+
+    resource->id = id;
+    resource->type = SOL_LWM2M_RESOURCE_TYPE_MULTIPLE;
+    resource->data_type = data_type;
+    resource->data = calloc(resource_len, sizeof(struct sol_lwm2m_resource_data));
+    SOL_NULL_CHECK(resource->data, -ENOMEM);
+    resource->data_len = resource_len;
+
+    for (i = 0; i < resource_len; i++) {
+        void *v = sol_vector_get_no_check(res_inst_ids, i);
+        uint16_t *ui = v;
+        resource->data[i].id = *ui;
+
+        v = sol_vector_get_no_check(res_inst_values, i);
+
+        if (resource->data_type == SOL_LWM2M_RESOURCE_DATA_TYPE_OPAQUE ||
+            resource->data_type == SOL_LWM2M_RESOURCE_DATA_TYPE_STRING) {
+            struct sol_blob *blob = v;
+            SOL_NULL_CHECK_GOTO(blob, err_exit);
+            resource->data[i].content.blob = sol_blob_ref(blob);
+            SOL_NULL_CHECK_GOTO(resource->data[i].content.blob, err_ref);
+
+        } else if (resource->data_type == SOL_LWM2M_RESOURCE_DATA_TYPE_FLOAT) {
+            double *d = v;
+            resource->data[i].content.fp = *d;
+
+        } else if (resource->data_type == SOL_LWM2M_RESOURCE_DATA_TYPE_INT ||
+            resource->data_type == SOL_LWM2M_RESOURCE_DATA_TYPE_TIME) {
+            int64_t *isf = v;
+            resource->data[i].content.integer = *isf;
+
+        } else if (resource->data_type == SOL_LWM2M_RESOURCE_DATA_TYPE_BOOL) {
+            int *ic = v;
+            resource->data[i].content.integer = *ic;
+
+        } else if (resource->data_type == SOL_LWM2M_RESOURCE_DATA_TYPE_OBJ_LINK) {
+            int *ic = v;
+            resource->data[i].content.integer = (uint16_t)*ic;
+            resource->data[i].content.integer = (resource->data[i].content.integer << 16) |
+                (uint16_t)*ic;
+
+        } else {
```

Comment:
They are right there:
```C
        } else if (resource->data_type == SOL_LWM2M_RESOURCE_DATA_TYPE_INT ||
            resource->data_type == SOL_LWM2M_RESOURCE_DATA_TYPE_TIME) {
```

I'm not sure an auxiliary function would help that much, since there are differences between this `if-else` chain and the one in `sol_lwm2m_resource_init` (like the pointer variable declaration depending on the `resource->data_type` and the use [or not] of `va_args`).